### PR TITLE
[train] Disallow trainer_resources() in ScalingConfig for GBDT trainers

### LINKED ISSUE #39058: [train] XGBoostTrainer silently disregards `ScalingConfig.trainer_resources`
### What happened + What you expected to happen

XGBoostTrainer (and LightGBMTrainer) use the XGBoost-Ray/LightGBM-Ray specific `get_tune_resources()` API to create a placement group factory:

https://github.com/ray-project/ray/blob/master/python/ray/train/gbdt_trainer.py#L363-L365

This will never schedule any resources for the trainer instance. In most cases, this is desired behavior.

However, the trainers still accept a `ScalingConfig.trainer_resources` argument and silently drop its contents. Instead, we should raise an error if the setting is set.

### Versions / Dependencies

master

### Reproduction script

```
from typing import Tuple

import ray
from ray import tune
from ray.data import Dataset
from ray.train.xgboost import XGBoostTrainer
from ray.train import Result, ScalingConfig


def prepare_data() -> Tuple[Dataset, Dataset]:
    dataset = ray.data.read_csv("s3://anonymous@air-example-data/breast_cancer.csv")
    train_dataset, valid_dataset = dataset.train_test_split(test_size=0.3)
    return train_dataset, valid_dataset


def train_xgboost(num_workers: int, use_gpu: bool = False) -> Result:
    train_dataset, valid_dataset = prepare_data()

    # XGBoost specific params
    params = {
        "tree_method": "approx",
        "objective": "binary:logistic",
        "eval_metric": ["logloss", "error"],
    }

    trainer = XGBoostTrainer(
        scaling_config=ScalingConfig(num_workers=num_workers, use_gpu=use_gpu, trainer_resources={"CPU": 5000}),
        label_column="target",
        params=params,
        datasets={"train": train_dataset, "valid": valid_dataset},
        num_boost_round=100,
    )
    tuner = tune.Tuner(trainer)
    result = tuner.fit()[0]
    print(result.metrics)

    return result


ray.init(num_cpus=3)
result = train_xgboost(num_workers=2, use_gpu=False)
```

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

`ScalingConfig.trainer_resources` are currently silently dropped in GBDT trainers. Instead, we should raise an error when this is set.

## Related issue number

Closes #39058

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/_internal/config.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ def ensure_only_allowed_dataclass_keys_updated(
         key
         for key in prohibited_keys
         if dataclass.__dict__[key] != default_data.__dict__[key]
+        and (dataclass.__dict__[key] or default_data.__dict__[key])
```
Comment (matthewdeng) [inline python/ray/air/_internal/config.py:42]: Why is this needed?

File: python/ray/train/gbdt_trainer.py [hunk 2]
```diff
@@ -128,7 +128,8 @@ class GBDTTrainer(BaseTrainer):
         **train_kwargs: Additional kwargs passed to framework ``train()`` function.
     """
 
-    _scaling_config_allowed_keys = BaseTrainer._scaling_config_allowed_keys + [
+    _scaling_config_allowed_keys = [
+        "_max_cpu_fraction_per_node",
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/train/gbdt_trainer.py:132]: Enabling it requires changes in XGBoost-Ray - we can do this, but I'd suggest we first land this (currently it's silently ignoring the setting) and revise this later. Maybe as part of merging the XGBoost-Ray repo into Ray Train.

[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/train/gbdt_trainer.py:132]: Sounds good could you add a TODO in this PR? I want to make sure this is captured explicitly since it's a bit unintuitive without this context.

File: python/ray/air/config.py [hunk 3]
```diff
@@ -259,7 +259,7 @@ def from_placement_group_factory(
     ) -> "ScalingConfig":
         """Create a ScalingConfig from a Tune's PlacementGroupFactory"""
         if pgf.head_bundle_is_empty:
-            trainer_resources = {}
+            trainer_resources = None
```
Comment (matthewdeng) [inline python/ray/air/config.py:262]: Hm I'm not fully familiar with the logic here, but are there other places in the code that might have expected this to be an empty dictionary?

Comment (krfricke) [reply python/ray/air/config.py:262]: Actually, yeah, I believe this changes the logic and has to be reverted.

Comment (krfricke) [reply python/ray/air/config.py]: I've moved the check.

File: python/ray/train/gbdt_trainer.py
Comment (matthewdeng) [inline python/ray/train/gbdt_trainer.py:1]: Could you update documentation to reflect this? 2 places this comes up:
* https://github.com/ray-project/ray/blob/5b754286201dd56f70b14e57e1011c5c3be44965/doc/source/train/distributed-xgboost-lightgbm.rst?plain=1#L86
* The docstrings for the different GBDT trainers, since `ScalingConfig` says: https://github.com/ray-project/ray/blob/5b754286201dd56f70b14e57e1011c5c3be44965/python/ray/air/config.py#L97-L98

**Review by matthewdeng — APPROVED** [review]
